### PR TITLE
usability improvements

### PR DESCRIPTION
**Pull request by revintec**
- [ ] This change has been discussed with the author.
- [ ] It implements / fixes issue #.

---
sorry for not opening an issue in advance discussing these changes
if these changes are unacceptable, please reply and I'll change them accordingly

1) MAJOR usability fix: text unreadable under OSX 10.10 dark mode
image
2) add a new UI OSC layout: "Always"
  2.1) always showing a minimal UI, won't be hidden automatically but only through user interaction
  2.2) more transparent UI elements, and more simpler controls to prevent occlusion of playback**
  2.3) take playback speed into account when calculating remaining time
3) changes made to OSC layout "Bottom" because of the "Always" layout
  3.1) shrink height to prevent occlusion of subtitles in "Always" mode, needs more work*
  3.2) change timestring position during hover seek because of the shrinking, needs more work*

** I originally wrote a QuickTime Player plugin, among many features are:
image
image
1) simplified OSC to always keep track of media playback progress
2) OSC is more transparent to minimize occlusion of playback
3) real-world time tracking of remaining time, because I always speed up playback to save time
4) realtime seek using trackpad ranging from single frame precision to fast overview automatically

these I favors the most and want to port back to IINA A.S.A.P, however my development environment is very constrained, I'm using OSX 10.10 which made me unable to view and change code in my machine, so I have to do these in a very sluggish virtual machine and this slows down my development dramastically
\* ideally changes I made shouldn't interfere with other code like "Bottom" layout, but it was too difficult for me to code in a virtual machine, so I'll do it in a later time or maybe you can help me do this?

### CODE REVIEW

File: iina/DurationDisplayTextField.swift [hunk 1]
```diff
@@ -29,7 +29,7 @@ class DurationDisplayTextField: NSTextField {
   }
   
   
-  func updateText(with duration: VideoTime, given current: VideoTime) {
+  func updateText(with duration: VideoTime, given current: VideoTime, speed of: Double) {
```
Comment (saagarjha) [inline iina/DurationDisplayTextField.swift]: The new parameter label doesn't really fit in with the rest of the function declaration. Could you try rephrasing this?

File: iina/MainWindowController.swift [hunk 2]
```diff
@@ -730,13 +731,13 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
       oscTopMainView.addView(fragSliderView, in: .leading)
       oscTopMainView.setClippingResistancePriority(.defaultLow, for: .horizontal)
       oscTopMainView.setVisibilityPriority(.detachOnlyIfNecessary, for: fragVolumeView)
-    case .bottom:
+    case .bottom: fallthrough case .always:
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Please put this on two lines.

File: iina/MainWindowController.swift [hunk 3]
```diff
@@ -965,7 +966,7 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
       isMouseInWindow = false
       if controlBarFloating.isDragging { return }
       destroyTimer()
-      hideUI()
+      if oscPosition != .always { hideUI() }
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: I'd prefer if this was on three lines

File: iina/MainWindowController.swift [hunk 4]
```diff
@@ -1756,7 +1753,7 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
 
     case .ultraDark:
       appearance = NSAppearance(named: .vibrantDark)
-      material = .ultraDark
+      if #available(OSX 10.11, *) { material = .ultraDark } else { material = .dark }
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Put this on five lines please

File: iina/MainWindowController.swift [hunk 5]
```diff
@@ -1766,7 +1763,7 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
 
     case .mediumLight:
       appearance = NSAppearance(named: .vibrantLight)
-      material = .mediumLight
+      if #available(OSX 10.11, *) { material = .mediumLight } else { material = .light }
```
Comment (saagarjha) [inline iina/MainWindowController.swift]: Likewise

File: iina/MiniPlayerWindowController.swift [hunk 6]
```diff
@@ -163,7 +163,8 @@ class MiniPlayerWindowController: NSWindowController, NSWindowDelegate {
     }
     let percentage = (pos.second / duration.second) * 100
     leftLabel.stringValue = pos.stringRepresentation
-    rightLabel.updateText(with: duration, given: pos)
+    // let speed = oscPosition != .always ? 0 : player.info.playSpeed
+    rightLabel.updateText(with: duration, given: pos, speed: /*speed*/0)
```
Comment (saagarjha) [inline iina/MiniPlayerWindowController.swift]: The comments here aren't necessary

File: iina/PrefUIViewController.swift [hunk 7]
```diff
@@ -50,7 +50,7 @@ class PrefUIViewController: NSViewController, MASPreferencesViewController {
       name = "osc_float"
     case 1:
       name = "osc_top"
-    case 2:
+    case 2: fallthrough case 3:
```
Comment (saagarjha) [inline iina/PrefUIViewController.swift]: Two lines

File: iina/DurationDisplayTextField.swift (hunk 1 above)
Comment (revintec) [reply iina/DurationDisplayTextField.swift]: I basically always uses positional parameters, so sorry for the bad taste of label selection
I've changed it to "and speed", is that OK?

File: iina/MainWindowController.swift (hunk 2 above)
Comment (revintec) [reply iina/MainWindowController.swift]: updated

File: iina/MainWindowController.swift (hunk 3 above)
Comment (revintec) [reply iina/MainWindowController.swift]: updated

File: iina/MainWindowController.swift (hunk 4 above)
Comment (revintec) [reply iina/MainWindowController.swift]: updated

File: iina/MainWindowController.swift (hunk 5 above)
Comment (revintec) [reply iina/MainWindowController.swift]: updated

File: iina/MiniPlayerWindowController.swift (hunk 6 above)
Comment (revintec) [reply iina/MiniPlayerWindowController.swift]: removed

File: iina/PrefUIViewController.swift (hunk 7 above)
Comment (revintec) [reply iina/PrefUIViewController.swift]: updated

**Comment by revintec** [comment]
sorry for the late reply, busy working during daytime
I've changed my commit accordingly, and replied all your comments
please check if there're other problems, thx in advance

**Comment by lhc70000** [comment]
Hi @revintec, thanks for your work and I'll review the code and discuss more about UI/UX with you tomorrow. My initial advice before taking a deeper look is, please make separate PR for each of the three points. We may need more discussion on the OSC layout.

**Comment by revintec** [comment]
@lhc70000 Hi, thx for the reply, I'm also feeling that a more thorough discussion is needed, my current UI/UX changes are rough. the 1st point is purely a bugfix, and can be separated effortlessly, but the last two points are a little intertwined in code, do you want me to separate this time or it's OK for me to do it next time?

**Comment by lhc70000** [comment]
@revintec Sorry for late reply; yes, please open a new PR for the bug fix.

**Comment by revintec** [comment]
@lhc70000 no problem. I'm a little busy right now, I'll do it later and open a new PR
also, I'll refine the UI related code and open a new PR, too. closing this one

**Pull request closed without merging**